doc/PROGMEM.rst

```diff
@@ -0,0 +1,266 @@
+Guide to PROGMEM on ESP8266 and Arduino IDE
+===========================================
+
+Intro
+-----
+
+On low memory devices like the arduino and esp8266 you do not want
+strings to be stored in RAM. This occurs by default on these systems.
+Declare a string ``const char * xyz = "this is a string"`` and it will
+use up RAM.
+
+The solution on these devices is to allow strings to be stored in read
+only memory, in Arduino this is the PROGMEM macro. Most of my experience
+is with the ESP8266 which is a 32bit micros controller. This device
+stores PROGMEM data in flash. The macro PROGMEM on ESP8266 is simply
+
+.. code:: cpp
+
+    #define PROGMEM   ICACHE_RODATA_ATTR
+
+Which in turn is defined by:
+
+.. code:: cpp
+
+    #define ICACHE_RODATA_ATTR  __attribute__((section(".irom.text")))
+
+Which places the variable in the .irom.text section ie flash.
+
+| The key to understanding PROGMEM is to understand how the strings are
+  stored and then how they are retrieved from flash.
+| ### Declare a global string to be stored in flash.
+
+.. code:: cpp
+
+    static const char xyz[] PROGMEM = "This is a string stored in flash";
+
+Declare a flash string within code block.
+-----------------------------------------
+
+For this you can use the PSTR macro. Which are all defined in
+`pgmspace.h <https://github.com/esp8266/Arduino/blob/master/cores/esp8266/pgmspace.h>`__
+
+.. code:: cpp
+
+    #define PGM_P       const char *
+    #define PGM_VOID_P  const void *
+    #define PSTR(s) (__extension__({static const char __c[] PROGMEM = (s); &__c[0];}))
+
+In practice:
+
+.. code:: cpp
+
+    void myfunction(void) {
+    PGM_P xyz = PSTR("Store this string in flash");
+    const char * abc = PSTR("Also Store this string in flash");
+    }
+
+The two examples above will store these strings in flash. To retrieve
+and manipulate flash strings is not straight forward as the esp8266 must
+read from flash in 4byte words. In the Arduino IDE for esp8266 there are
+several functions that can help retrieve strings from flash that have
+been stored using PROGMEM. Both of the examples above will give you a
+``const char *`` back, however if you try to do anything with these
+pointers, without correct 32bit alignment you will get a seg fault and
+the esp will crash. You must read from the flash 32 bit aligned.
+
+Functions to read back from PROGMEM
+-----------------------------------
+
+Which are all defined in
+`pgmspace.h <https://github.com/esp8266/Arduino/blob/master/cores/esp8266/pgmspace.h>`__
+
+.. code:: cpp
+
+    int memcmp_P(const void* buf1, PGM_VOID_P buf2P, size_t size);
+    void* memccpy_P(void* dest, PGM_VOID_P src, int c, size_t count);
+    void* memmem_P(const void* buf, size_t bufSize, PGM_VOID_P findP, size_t findPSize);
+    void* memcpy_P(void* dest, PGM_VOID_P src, size_t count);
+    char* strncpy_P(char* dest, PGM_P src, size_t size);
+    #define strcpy_P(dest, src)          strncpy_P((dest), (src), SIZE_IRRELEVANT)
```

Comment:
I would suggest to simplify this a bit by listing
```
char* strcpy_P(char* dest, PGM_P src);
```
instead of a macro (same for other macros). Makes this list easier to read.